engine/internal/processor/processor.go

```diff
@@ -89,47 +89,79 @@ type P struct {
 }
 
 // Run runs the processor.
+//
+// TODO(kenji): Gracefully handle an error from the server.
 func (p *P) Run(ctx context.Context) error {
 	ctx = auth.AppendWorkerAuthorization(ctx)
 	stream, err := p.client.ProcessTasks(ctx)
 	if err != nil {
 		return err
 	}
 
-	log.Printf("Registering engine: %s\n", p.engineID)
-	if err := p.sendEngineStatus(ctx, stream); err != nil {
-		return err
-	}
-	log.Printf("Registered engine: %s\n", p.engineID)
+	errCh := make(chan error, 1)
+	go func() {
+		if err := p.sendEngineStatusPeriodicaly(ctx, stream); err != nil {
+			errCh <- fmt.Errorf("send engine status: %s", err)
+		}
+	}()
 
 	go func() {
-		// Periodically send engine status.
-		for {
-			if err := p.sendEngineStatus(ctx, stream); err != nil {
-				log.Fatalf("Failed to send engine status: %s\n", err)
-			}
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(statusReportInterval):
-			}
+		if err := p.processTasks(ctx, stream, errCh); err != nil {
+			errCh <- fmt.Errorf("process tasks: %s", err)
 		}
 	}()
+	return <-errCh
+}
+
+func (p *P) sendEngineStatusPeriodicaly(
+	ctx context.Context,
+	stream v1.InferenceWorkerService_ProcessTasksClient,
+) error {
+	var isFirst bool
 
+	log.Printf("Registering engine: %s\n", p.engineID)
+	for {
+		if err := p.sendEngineStatus(ctx, stream); err != nil {
+			return err
+		}
+
+		if isFirst {
+			isFirst = false
+			log.Printf("Registered engine: %s\n", p.engineID)
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(statusReportInterval):
+		}
+	}
+}
+
+func (p *P) processTasks(
+	ctx context.Context,
+	stream v1.InferenceWorkerService_ProcessTasksClient,
+	errCh chan<- error,
+) error {
 	for {
 		resp, err := stream.Recv()
 		if err == io.EOF {
-			return nil
+			return fmt.Errorf("connection closed")
 		}
 		if err != nil {
-			return err
+			return fmt.Errorf("receive task: %s", err)
 		}
 
-		log.Printf("Started processing task: %s\n", resp.NewTask.Id)
-		if err := p.processTask(ctx, stream, resp.NewTask); err != nil {
-			return err
-		}
-		log.Printf("Completed task: %s\n", resp.NewTask.Id)
+		// Create a goroutine to process the task so that we can receive the
+		// next task. Ollama then might process requests in parallel.
```

Comment:
question: no needs to limit how many tasks process in parallel?

Reply:
Good point! We might introduce the limit later, but decided not to do now. The reason is to keep the logic simple. If we want to limit the number of requests in parallel, `inference-manager-server` should do so that it can distribute requests to other `engine` instance. Also vLLM can have its queue, so we don't need to have yet another queue here.